homeassistant/components/switch/__init__.py

```diff
@@ -69,38 +70,49 @@ def toggle(hass, entity_id=None):
     hass.services.call(DOMAIN, SERVICE_TOGGLE, data)
 
 
-def setup(hass, config):
+@asyncio.coroutine
+def async_setup(hass, config):
     """Track states and offer events for switches."""
     component = EntityComponent(
         _LOGGER, DOMAIN, hass, SCAN_INTERVAL, GROUP_NAME_ALL_SWITCHES)
-    component.setup(config)
+    yield from component.async_setup(config)
 
-    def handle_switch_service(service):
+    @asyncio.coroutine
+    def async_handle_switch_service(service):
         """Handle calls to the switch services."""
-        target_switches = component.extract_from_service(service)
+        target_switches = component.async_extract_from_service(service)
 
+        update_tasks = []
         for switch in target_switches:
             if service.service == SERVICE_TURN_ON:
-                switch.turn_on()
+                yield from switch.async_turn_on()
             elif service.service == SERVICE_TOGGLE:
-                switch.toggle()
+                yield from switch.async_toggle()
             else:
-                switch.turn_off()
+                yield from switch.async_turn_off()
 
             if switch.should_poll:
-                switch.update_ha_state(True)
-
-    descriptions = load_yaml_config_file(
-        os.path.join(os.path.dirname(__file__), 'services.yaml'))
-    hass.services.register(DOMAIN, SERVICE_TURN_OFF, handle_switch_service,
-                           descriptions.get(SERVICE_TURN_OFF),
-                           schema=SWITCH_SERVICE_SCHEMA)
-    hass.services.register(DOMAIN, SERVICE_TURN_ON, handle_switch_service,
-                           descriptions.get(SERVICE_TURN_ON),
-                           schema=SWITCH_SERVICE_SCHEMA)
-    hass.services.register(DOMAIN, SERVICE_TOGGLE, handle_switch_service,
-                           descriptions.get(SERVICE_TOGGLE),
-                           schema=SWITCH_SERVICE_SCHEMA)
+                update_coro = switch.async_update_ha_state(True)
+            else:
+                update_coro = switch.async_update_ha_state()
```

Comment:
We should not force an update of the state if we are not supposed to poll. `should_poll` being `False` means the platform itself is responsible for updating the state.